plutus-core/src/Language/PlutusCore/Parsable.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE DefaultSignatures    #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE OverloadedStrings    #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+
+{- | A typeclass which provides a `parseConstant` method to convert Text strings
+into objects of the appropriate type. This allows one to define parsers for
+literal constants belonging to extensible built-in types without having to
+modify the main Plutus Core parser. We expect parseConstant . prettyConst to be
+the identity function. -}
+
+module Language.PlutusCore.Parsable
+    (
+     Parsable(..)
+    )
+where
+
+import           PlutusPrelude
+
+import           Data.Bits            (shiftL, (.|.))
+import           Data.ByteString.Lazy (ByteString)
+import qualified Data.ByteString.Lazy as BSL (pack)
+import           Data.Char            (ord)
+import qualified Data.Text            as T
+import           Text.Read            (readMaybe)
+
+
+class Parsable a
+  where
+  parseConstant :: T.Text -> Maybe a
+  -- ^ Return Nothing if the string is invalid, otherwise the corresponding value of type a.
+  default parseConstant :: Read a => T.Text -> Maybe a
+  parseConstant = readMaybe @ a . T.unpack
+  -- ^ The default implmentation is 'read'
+
+instance Parsable Bool
+instance Parsable Char
+instance Parsable Integer
+instance Parsable String
+instance Parsable ()
+
+instance Parsable ByteString
+  where parseConstant = parseByteStringConstant
+
+
+--- Parsing bytestrings ---
+
+-- A literal bytestring consists of the '#' character followed immediately by
+-- an even number of upper- or lower-case hex digits.
+parseByteStringConstant :: T.Text -> Maybe ByteString
+parseByteStringConstant lit = do
+    case T.unpack lit of
+        '#':body -> asBSLiteral body
+        _        -> Nothing
+
+-- | Convert a list to a list of pairs, failing if the input list has an odd number of elements
+pairs :: [a] -> Maybe [(a,a)]
+pairs []         = Just []
+pairs (a:b:rest) = fmap ((:) (a,b)) (pairs rest)
```

Comment:
Total nitpicking: I'd perhaps save a few characters and write `((a,b) :) <$> pairs rest`, but obviously not important.

Reply:
OK, I'll do that.  I think it saves a whole four characters!

I'm still not quite sure where to put this file.  It's kind of dual to `PrettyConst`, so I was almost tempted to put it in there.  It should probably go somewhere near it though: if we have, say `true` and `false` for booleans then that should be defined in a single place where both `parseConstant` and `prettyConst` can refer to it.